internal/events/stream/event_test.go

```diff
@@ -0,0 +1,60 @@
+package stream
+
+import (
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	eventstream "github.com/openshift/assisted-service/pkg/stream"
+)
+
+type TestStruct struct {
+	Foo string `json:"foo"`
+	Bar string `json:"bar"`
+}
+
+var _ = Describe("Stream event", func() {
+	var (
+		stream *eventstream.MockEventStreamWriter
+		ctrl   *gomock.Controller
+	)
+	BeforeEach(func() {
+		ctrl = gomock.NewController(GinkgoT())
+		stream = eventstream.NewMockEventStreamWriter(ctrl)
+	})
+
+	It("return error if stream is not configured", func() {
+
+		res := StreamEvent(nil, "TestEvent", "mykey", nil)
+		Expect(res).ShouldNot(BeNil())
+	})
+
+	It("sends event with regular payload", func() {
+
+		payload := &TestStruct{
+			Foo: "bar",
+			Bar: "foo",
+		}
+		expectedPayload := []byte(`{"name":"TestEvent","payload":{"foo":"bar","bar":"foo"},"metadata":{"versions":{"assisted-installer":"quay.io/edge-infrastructure/assisted-installer:latest","assisted-installer-controller":"quay.io/edge-infrastructure/assisted-installer-controller:latest","assisted-installer-service":"Unknown","discovery-agent":"quay.io/edge-infrastructure/assisted-installer-agent:latest"}}}`)
```

Comment:
can we indent it for readability?
maybe even construct it as a map and marshal afterwards

Reply:
Although I would love to have better reliability here, I am afraid if we serialize it from a map it would lose the order.
If we populate the same data structure that we have in the productive code, we would be just replicating the same code in the test...
I'll try to think of something to improve readability but if you have other ideas please let me know.

Reply:
too bad golang doesn't allow to preserve order in json marshaling

Reply:
If you really need to preserve order in JSON documents you could use this library (or any other similar one):

https://gitlab.com/c0b/go-ordered-json

But I'd say that in this context it would be preferable to compare the expected and actual JSON documents ignoring order. For that I'd suggest to create a new gomock matcher similar to the gomega `MatchJSON`:

https://onsi.github.io/gomega/#matchjsonjson-interface

Could be something like this:

```go
func JSON(expected string) gomock.Matcher {
	var expectedAny any
	err := json.Unmarshal([]byte(expected), &expectedAny)
	if err != nil {
		panic(err)
	}
	buffer := &bytes.Buffer{}
	encoder := json.NewEncoder(buffer)
	encoder.SetIndent("", "\t")
	err = encoder.Encode(expectedAny)
	if err != nil {
		panic(err)
	}
	expectedText := buffer.String()
	return &jsonMatcher{
		expectedText: expectedText,
		expectedAny:  expectedAny,
	}
}

type jsonMatcher struct {
	expectedText string
	expectedAny  any
}

func (m *jsonMatcher) Matches(actual any) bool {
	actualText := actual.(string)
	var actualAny any
	err := json.Unmarshal([]byte(actualText), &actualAny)
	if err != nil {
		panic(err)
	}
	return reflect.DeepEqual(actualAny, m.expectedAny)
}

func (m *jsonMatcher) String() string {
	return m.expectedText
}
```

Then you should be able to do something like this in your test:

```go
stream.EXPECT().Write(
        []byte("mykey"),
        JSON(`{
                "name": "TestEvent",
                "payload": ...
        }`),
)
```

The matcher will ignore what isn't relevant in a JSON document, like extra space and ordering of fields.

Reply:
That is a great idea @jhernand